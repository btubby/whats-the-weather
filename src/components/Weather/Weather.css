
.weather {
  background-color: #492F46;
  padding: 30px;
  text-align: left;
  font-size: 1.2rem;
}

.left {
  float: left;
}
.left2 {
  overflow: hidden;
  display: block;
  padding: 0 4px 0 10px
}
.left2 input {
  width: 100%
}

.wrapper {
  display:flex;
  height: 250px;
}
.refresh {
  width: 100%;
  font-size: 1rem;
  text-align: right;
}
.search-history {
  /* border: 1px solid green; */
  padding-left: 3px;
}