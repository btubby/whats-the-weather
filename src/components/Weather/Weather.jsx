import React, { useState, useEffect, useRef } from "react";
import { WeatherResult } from "../WeatherResult/WeatherResult";
import "./Weather.css";

export const Weather = (props) => {
  const refreshTimeoutSeconds = 60;

  const [location, setLocation] = useState("");
  const [hasError, setError] = useState(false);
  const [apiResponse, setApiResponse] = useState();
  const [seconds, setSeconds] = useState(refreshTimeoutSeconds);
  const isMounted = useRef(false);

  // Side effect to set (and clear) a new timer each time a valid location is entered
  // Don't set timer on component mount
  useEffect(() => {
    if (isMounted.current && !hasError) {
      console.log("fetching");
      const interval = setInterval(() => {
        setSeconds((seconds) => seconds - 1);
      }, 1000);
      return () => clearInterval(interval);
    } else {
      isMounted.current = true;
    }
  }, [location, apiResponse]);

  // Side effect to make a new API call when the minute timer is up
  useEffect(() => {
    if (seconds <= 0) {
      resetInterval();
      fetchData();
    }
  }, [seconds]);
  const handleSubmit = (event) => {
    event.preventDefault();
    fetchData();
    resetInterval();
  };

  const resetInterval = () => {
    setSeconds(refreshTimeoutSeconds);
  };

  const fetchData = () => {
    fetch(
      `https://api.openweathermap.org/data/2.5/weather?q=${location}&appid=${props.apiKey}&units=metric`
    )
      .then(function (response) {
        return response.json();
      })
      .then(function (json) {
        json.cod = 404 && setError(json.message); // city not found
        json.cod = 400 && setError(json.message); // Nothing to geocode
        json.cod = 200 && setApiResponse(json);
      })
      .catch((error) => {
        setError(error);
      });
  };

  return (
    <div className="weather">
      <form onSubmit={handleSubmit}>

        <label className="left" htmlFor="location">
          Enter your location:&nbsp;
        </label>

        <span className="left2">
        <input
          id="location"
          type="text"
          value={location}
          onChange={(e) => setLocation(e.target.value)}
          autofocus="true"
        />
        </span>

      </form>
      {hasError && <h3>{hasError}</h3>}

      {apiResponse && apiResponse.main && (
        <WeatherResult response={apiResponse} />
      )}
      {apiResponse && !hasError &&
        <div className="refresh">
          Refreshing in {seconds}s
        </div>
      }
      </div>
  );
};
