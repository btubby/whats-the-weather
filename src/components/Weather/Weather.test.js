import React from 'react';
import { Weather } from './Weather';
import { unmountComponentAtNode, render } from 'react-dom';
import { act } from 'react-dom/test-utils';

let container = null;
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe("Given a Weather component", () => {
  it('renders correctly', () => {
    act(() => {
      render(<Weather />, container);
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});

