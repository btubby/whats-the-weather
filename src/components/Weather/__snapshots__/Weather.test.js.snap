// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a Weather component renders correctly 1`] = `
<div
  class="wrapper"
>
  <div
    class="weather"
  >
    <form>
      <label
        class="left"
        for="location"
      >
        Enter your location:Â 
      </label>
      <span
        class="left2"
      >
        <input
          id="location"
          type="text"
          value=""
        />
      </span>
      <input
        id="temperature-unit"
        type="checkbox"
      />
      <label
        class="display-kelvin"
        for="temperature-unit"
      >
        Display temperatures in Kelvin
      </label>
    </form>
  </div>
  <div
    class="search-history"
  />
</div>
`;
